File: api/src/tasks/borderTax/states/uk.ts
import { normalizeISODate, dateParts } from "../shared/dates";

/**
 * Uttarakhand border-tax AI prompt (fallback path).
 *
 * The scripted runner (worker/src/scripted/border_tax/uk.py) is the primary
 * path. This prompt is used when the scripted runner is disabled or hands off
 * to the AI agent.
 *
 * UK is NET-BANKING ONLY — the portal's payment gateway offers a single option,
 * "SBI (Multi Bank Payment)". There is no UPI / QR flow. The agent therefore
 * fills the form through the SBI gateway selection, then hands the actual
 * payment to a human (wait_for_human) and polls for the receipt afterwards.
 */
export const buildPrompt = async (p: Record<string, string>): Promise<string> => {
    const vehicleNumber = p.vehicleNumber;
    const taxMode = p.taxMode || "DAYS";
    const taxFrom = p.taxFrom;
    const taxUpto = p.taxUpto;

    const taxFromISO = normalizeISODate(taxFrom!);
    const taxUptoISO = normalizeISODate(taxUpto!);
    const tf = dateParts(taxFromISO);
    const tu = dateParts(taxUptoISO);

    const entryDistrict = p.entryDistrict || "DEHRADUN";
    // UK checkpost names (ASHARODI / KULHAL / TIMLI / TUNI) don't track the
    // district name, so we let the agent pick the FIRST available option.
    const entryCheckpoint = p.entryCheckpoint || "";
    const serviceType = p.serviceType || "Air Conditioned Service";
    const permitType = p.permitType || "TEMPORARY PERMIT";

    return `
You are a strict automation agent paying border tax for vehicle ${vehicleNumber} entering UTTARAKHAND.
You follow the steps below EXACTLY. You do NOT improvise, explore, or try alternative approaches.
Payment method: SBI (Multi Bank Payment) — net banking. There is NO UPI option on this portal.

===
IDENTITY & BEHAVIOR
===
- You are an instruction-follower, NOT a problem-solver.
- You execute ONLY the steps listed below, in the EXACT order listed.
- You NEVER click buttons, links, or UI elements not explicitly mentioned here.
- You NEVER navigate to URLs not explicitly listed here.
- You NEVER use JavaScript, console, evaluate(), or any programmatic scraping.

===
TOOLS
===
- wait_for_human → Call ONLY when explicitly told to in the steps below (CAPTCHA help if needed, and the net-banking payment step).
- save_receipt → Call ONCE on the final receipt page (see Phase 8).

===
STRICTLY FORBIDDEN ACTIONS
===
1. Using JavaScript evaluate() or console commands.
2. Navigating to any URL not listed in these instructions.
3. Clicking any element not mentioned in these instructions.
4. Submitting any form not described here.
5. Clicking the "Print" or "Back" button on the receipt page.

===
PHASE 1 — NAVIGATE TO CHECKPOST PORTAL
===
1. Go to https://parivahan.gov.in/en/node/579
2. In the 'Checkpost Tax' dropdown (placeholder '--- Select State Name ---'), select "UTTARAKHAND".
3. This navigates to the Online Checkpost Portal page.

===
PHASE 2 — SELECT SERVICE AND ENTER VEHICLE
===
1. Click the "Service Name" dropdown and select "VEHICLE TAX COLLECTION (OTHER STATE)".
2. Click the green ">> Go" button.
3. On the next page, type "${vehicleNumber}" into the "Input Vehicle Number" field.
4. Click the "Get Details" button.
5. Wait for the owner/vehicle details to load below.
   - If an error appears or no details load → ABORT. Reason: "Get Details failed: [exact error]".
   - If a "pending transaction" popup appears → ABORT. Reason: "Vehicle has a pending transaction".

===
PHASE 3 — ENTRY DETAILS
===
1. In the "Entry District Name" dropdown, select "${entryDistrict}".
   - If that exact district is not present, select the FIRST available district option.
2. Wait ~1s for the "Entry CheckPost Name" dropdown to populate, then select the FIRST available
   checkpost option${entryCheckpoint ? ` (prefer "${entryCheckpoint}" if present)` : ""}.
3. Click the "Next" button.

===
PHASE 4 — VEHICLE INFORMATION
===
1. Vehicle Category: if it is already filled, LEAVE IT. If empty, select the FIRST available option.
2. Permit Type: if it is already filled, LEAVE IT. If empty, select the FIRST available option
   (prefer "${permitType}" if present).
3. Service Type: select "${serviceType}" (the Air Conditioned option). This is the default for UK.
4. Click the "Next" button.

===
PHASE 5 — TAX INFORMATION
===
1. Tax Mode: select "${taxMode}".
2. Tax From: set the date field (id "floatingTaxfrom", type=date) to "${tf.iso}" (ISO YYYY-MM-DD).
   Verify it displays the chosen date, not the empty placeholder.
3. Tax Upto: set the date field (id "uptpDate", type=date) to "${tu.iso}" (ISO YYYY-MM-DD).
   Verify it displays the chosen date.
   - If either date refuses to stick, it is likely before the field's min date → ABORT.
     Reason: "Tax date before allowed minimum".
4. Click the "Calculate Fee/Tax" button. Wait for the tax table (MV Tax / Service-User Charge /
   Cess rows) and a Total Amount > 0 to appear.
5. Click the "Next" button.

===
PHASE 6 — DISCLAIMER (CAPTCHA + CONFIRM)
===
1. The Disclaimer page (Step 4 of 4) loads with the vehicle/tax summary.
2. Read the captcha image (a small canvas with distorted characters next to a blue refresh button)
   and type the characters exactly into the captcha input field (id "inputcap", case-sensitive,
   max 8 chars).
3. Tick the "I confirm that above information are correct as per my knowledge" checkbox.
   - If a popup appears (vehicle summary / "Receipt valid for X Days"), click its Close button.
4. Click the green "Pay Online" button at the bottom-right.
5. A confirmation popup appears: "Are you sure? You want to pay online ?". Click the green "Yes".
6. Wait for navigation to the Payment Gateway page (URL contains "etranspgi" or "paymentgateway").
   - If the captcha is rejected ("Invalid Captcha"), close the popup, click the blue refresh button,
     and retry with the new image. Max 3 retries, then ABORT.

===
PHASE 7 — PAYMENT GATEWAY
===
1. On the Payment Gateway page, open the "Select Payment" dropdown (id "dropOperator") and select
   "SBI (Multi Bank Payment)".
2. Tick the "I accept terms and conditions." checkbox.
3. Click the "Submit" button (id "sendSubmit").
4. Wait for redirect to the SBI multi-bank payment page.

===
PHASE 8 — HUMAN PAYMENT + RECEIPT CAPTURE
===
1. There is NO UPI/QR on UK. Call wait_for_human with reason:
   "Net-banking payment required for Uttarakhand border tax of vehicle ${vehicleNumber}. Please
   select your bank on the SBI Multi Bank Payment page, complete the payment (including OTP), and
   reply 'done' when finished."
2. After the human replies (or as soon as the page changes), wait for the Uttarakhand receipt page.
   Poll the page every few seconds (up to ~15 minutes). The receipt page is identified by ALL of:
     - "GOVERNMENT OF UTTARAKHAND"
     - "Checkpost Tax e-Receipt"
     - a non-empty "Receipt No." value
     - "Registration No." showing "${vehicleNumber}"
   Do NOT click "Click here" on any SBI success page — let the auto-redirect happen.
   - If a "Transaction Pending" / "Transaction Failed" message appears → ABORT.
     Reason: "Payment pending/failed at bank side".
3. Read from the receipt page: Receipt No. → receiptNumber; Grand Total number → amount;
   Payment Confirmation Date (or Payment Initialization Date) → paymentDate (convert to YYYY-MM-DD).
4. Call save_receipt EXACTLY ONCE:
   {"vehicleNumber":"${vehicleNumber}","receiptNumber":"<receiptNumber>","amount":<amount>,"paymentDate":"<YYYY-MM-DD>"}
   - If the response has "ok": true AND "pdfUploaded": true → COMPLETION (full success).
   - Otherwise → PARTIAL COMPLETION. Do NOT retry. Do NOT call save_receipt again.
   Do NOT click "Print" or "Back".

===
PARTIAL COMPLETION (payment likely succeeded, receipt capture failed)
===
Call done with:
  Vehicle: ${vehicleNumber}
  State: Uttarakhand
  Entry District: ${entryDistrict}
  Entry Checkpoint: ${entryCheckpoint || "<auto-selected>"}
  Permit Type: ${permitType}
  Service Type: ${serviceType}
  Tax Mode: ${taxMode}
  Tax Period: ${tf.iso} to ${tu.iso}
  Payment Method: SBI Net Banking (Multi Bank Payment)
  Amount Paid: ₹<amount if known, otherwise "unknown">
  Receipt Number: <receiptNumber if read, otherwise "unknown">
  Receipt PDF: not uploaded — <reason>
  Status: partial

===
COMPLETION (full success)
===
Reach this ONLY when save_receipt returned "ok": true AND "pdfUploaded": true. Call done with:
  Vehicle: ${vehicleNumber}
  State: Uttarakhand
  Entry District: ${entryDistrict}
  Entry Checkpoint: ${entryCheckpoint || "<auto-selected>"}
  Permit Type: ${permitType}
  Service Type: ${serviceType}
  Tax Mode: ${taxMode}
  Tax Period: ${tf.iso} to ${tu.iso}
  Payment Method: SBI Net Banking (Multi Bank Payment)
  Amount Paid: ₹<amount>
  Receipt Number: <receiptNumber>
  Receipt PDF: uploaded
  Status: complete
`.trim();
};

File: worker/src/scripted/border_tax/uk.py
# worker/src/scripted/border_tax/uk.py
"""Uttarakhand border-tax scripted runner.

NO AI. Pure scripted form-fill (parivahan Phases 1-5), then hand the
captcha + payment to a human and poll for the receipt in the background.
See `_handover_runner.run_handover_flow` for the full phase walk and the
handover semantics. This is now identical in shape to HP and BR — UK used
to be the AI-handoff variant, and this rewrite collapses it onto the shared
flow.

WHAT CHANGED vs the old UK runner
=================================
  - REMOVED the Phase 6 AI captcha handoff and the Phase 7 payment-gateway
    selection (SBI Multi Bank Payment). The shared flow stops right after
    Phase 5 (Calculate Fee/Tax → Next) on the Disclaimer page and hands the
    captcha + payment to the human, exactly like HP/BR. There is no AI for
    any source now ("app" and "web" behave identically).
  - REMOVED the pending-transaction auto-clear. A "pending transaction"
    popup after Get Details now ABORTS immediately with
    abort_reason="pending_transaction_popup" and the summary
    "Vehicle <n> already has a pending transaction at the parivahan portal.
     Please clear the pending transaction first, then retry." (NO auto-clear),
    matching HP/BR.

UK specifics (everything else comes from the shared flow + _STATE_DEFAULTS)
==========================================================================
  - Phase 1 state value: "UK".
  - Phase 3 Entry District default: DEHRADUN (via params/_STATE_DEFAULTS).
  - Phase 3 Entry Checkpost: UK's checkpost names (ASHARODI / KULHAL / TIMLI
    / TUNI) do NOT track the district name, so checkpost_strategy=
    "first_option" — pick the first available checkpost.
  - Phase 4: Vehicle Category / Permit Type (default TEMPORARY PERMIT,
    fallback ALL INDIA TOURIST PERMIT) / Service Type — "leave if already
    filled, set if empty". Service Type's default is "Air Conditioned
    Service", so an empty Service Type is set to AC; a pre-filled value is
    kept as-is (this is the HP/BR rule; the old runner force-set AC even when
    pre-filled). UK has NO Distance field (like PB/MP).
  - Phase 5 Tax Mode: RC-dependent (the portal renders different modes per
    vehicle). The requested mode is matched by <option> value; if the portal
    doesn't offer it for the RC, the run aborts with
    abort_reason="tax_mode_not_offered_for_rc" listing the modes that were
    offered. Tax From/Upto input types are auto-detected (date vs
    datetime-local); Tax Upto is filled only in DAYS mode.

NET-BANKING-ONLY NOTE
=====================
UK's portal only offers "SBI (Multi Bank Payment)" net banking — there is no
UPI / QR flow. The human completes the net-banking payment themselves at the
handover; we pass poll_qr=False and a 15-minute window (HANDOVER_TIMEOUT_SECS
in the shared runner) and poll the page for the receipt in the background.
The moment the receipt renders we capture the PDF and finish — we never block
on an explicit human "done", since they may pay and never notify us.
run_job.py treats UK as scripted-eligible even though paymentMethod != "upi"
(see scripted.runner.state_is_net_banking_scripted), so this runs scripted
regardless of source or paymentMethod.
"""

from __future__ import annotations

import re

from ..log import StepLogger
from ..types import RunOutcome
from .params import BorderTaxParams
from ._payment_wait import PaymentCaptureConfig
from ._handover_runner import StateHandoverConfig, run_handover_flow


# ─── Receipt parsing ───────────────────────────────────────────────────

_MONTH_TO_NUM = {
    "JAN": "01",
    "FEB": "02",
    "MAR": "03",
    "APR": "04",
    "MAY": "05",
    "JUN": "06",
    "JUL": "07",
    "AUG": "08",
    "SEP": "09",
    "OCT": "10",
    "NOV": "11",
    "DEC": "12",
}


def _normalize_receipt_date(s: str) -> str | None:
    """26-Apr-2026 -> 2026-04-26."""
    m = re.match(r"(\d{1,2})-(\w{3})-(\d{4})", s.strip(), re.IGNORECASE)
    if not m:
        return None
    day, mon_abbr, year = m.group(1).zfill(2), m.group(2).upper(), m.group(3)
    mon_num = _MONTH_TO_NUM.get(mon_abbr)
    if not mon_num:
        return None
    return f"{year}-{mon_num}-{day}"


async def _extract_receipt_fields(session, vehicle_number: str) -> dict | None:
    """Scrape receiptNumber, amount, paymentDate from the receipt page text.

    Generic regex set (same as HP/BR/PB) — no confirmed UK receipt-number
    prefix yet, so we accept any alphanumeric receipt number. Tighten once
    a real UK receipt sample is available.
    """
    from ..steps import _cdp_eval

    text = await _cdp_eval(session, "document.body.innerText || ''")
    if not text:
        return None

    receipt_match = re.search(
        r"Receipt\s*No\.?\s*:?\s*([A-Z0-9]+)",
        text,
        re.IGNORECASE,
    )
    receipt_number = receipt_match.group(1) if receipt_match else None

    amount_match = re.search(
        r"Grand\s*Total\s*:?\s*(\d+(?:\.\d+)?)",
        text,
        re.IGNORECASE,
    )
    amount = amount_match.group(1) if amount_match else None

    date_match = re.search(
        r"Payment\s*Confirmation\s*Date\s*:?\s*([\d]{1,2}-\w{3}-\d{4})",
        text,
        re.IGNORECASE,
    )
    payment_date = _normalize_receipt_date(date_match.group(1)) if date_match else None

    if not (receipt_number or amount):
        return None
    return {
        "receiptNumber": receipt_number,
        "amount": amount,
        "paymentDate": payment_date,
        "vehicleNumber": vehicle_number,
    }


# ─── Config ────────────────────────────────────────────────────────────

# poll_qr=False is passed inside run_handover_flow (UK has no UPI/QR), so
# qr_selector is unused but the dataclass field is required.
_UK_PAYMENT_CONFIG = PaymentCaptureConfig(
    state_name="Uttarakhand",
    qr_selector="img#qrcodeImg",  # unused (poll_qr=False at the call site)
    receipt_markers=[
        "GOVERNMENT OF UTTARAKHAND",
        "CHECKPOST TAX E-RECEIPT",
        "Receipt No",
        # "GRAND TOTAL",
    ],
    positive_markers_regex=[
        r"payment\s*successful",
        r"transaction\s*successful",
        r"successfully\s*paid",
        r"transaction\s*status\s*[:\-]?\s*success",
        r"government\s*of\s*uttarakhand",
        r"checkpost\s*tax\s*e-?receipt",
    ],
    negative_markers_regex=[
        r"transaction\s*status\s*[:\-]?\s*pending",
        r"your\s*transaction\s*status\s*is\s*pending",
        r"transaction\s*confirmation\s*pending",
        r"transaction\s*status\s*[:\-]?\s*failed",
        r"transaction\s*failed",
        r"payment\s*failed",
    ],
)

_UK_CONFIG = StateHandoverConfig(
    state_code="UK",
    state_name="Uttarakhand",
    checkpost_strategy="first_option",  # ASHARODI/KULHAL/TIMLI/TUNI ≠ district
    payment_config=_UK_PAYMENT_CONFIG,
    extract_receipt_fields=_extract_receipt_fields,
)


# ─── Public entrypoint ─────────────────────────────────────────────────


async def run(
    session,
    params: BorderTaxParams,
    log: StepLogger,
) -> RunOutcome:
    """Run UK scripted border-tax. Always returns a RunOutcome."""
    return await run_handover_flow(session, params, log, config=_UK_CONFIG)
